Initial work toward synthetic monitors

andrewvc opened the pull request:
Initial, experimental version of synthetic monitor support for Heartbeat. This works in concert with https://github.com/elastic/synthetics , and is intended to only be run within the docker container built in the synthetics repo. This is guarded by the new `ELASTIC_SYNTHETICS_CAPABLE` env variable which must be set to `true`. As a note, this is easy to bypass because the main purpose is to make it clear that we do not support synthetics outside of the docker container for now.

This includes a new directory, `sample-synthetics-config` in the x-pack directory that demonstrates using heartbeat with synthetics. In the future 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

You can check that this PR works by running (in the x-pack/heartbeat directory): 
`mage build && env ELASTIC_SYNTHETICS_CAPABLE=true ./heartbeat --path.config sample-synthetics-config/ -e`

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21436/detail/PR-21436/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21436/detail/PR-21436/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21436/detail/PR-21436/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-21436/detail/PR-21436/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_21436.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #21436 opened]`

* Start Time: 2020-09-30T19:30:44.067+0000

* Duration: 11 min 55 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `make check`
    * Description: make check
    
    * Duration: 4 min 40 sec
    
    * Start Time: 2020-09-30T19:37:00.425+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-21436/runs/1/steps/146/log)
    
  
  * **Name**: `Notifies GitHub of the status of a Pull Request`
    * Description: script returned exit code 2
    
    * Duration: 0 min 1 sec
    
    * Start Time: 2020-09-30T19:40:48.357+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-21436/runs/1/steps/182/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-09-30T19:40:40.763Z]   GitCommit:        fec3683
[2020-09-30T19:40:40.763Z] Unable to find image 'alpine:3.4' locally
[2020-09-30T19:40:41.336Z] 3.4: Pulling from library/alpine
[2020-09-30T19:40:41.597Z] c1e54eec4b57: Pulling fs layer
[2020-09-30T19:40:41.860Z] c1e54eec4b57: Download complete
[2020-09-30T19:40:42.123Z] c1e54eec4b57: Pull complete
[2020-09-30T19:40:42.123Z] Digest: sha256:b733d4a32c4da6a00a84df2ca32791bb03df95400243648d8c539e7b4cce329c
[2020-09-30T19:40:42.123Z] Status: Downloaded newer image for alpine:3.4
[2020-09-30T19:40:43.841Z] + python .ci/scripts/pre_archive_test.py
[2020-09-30T19:40:43.841Z] Copy ./filebeat/build into build/filebeat/build
[2020-09-30T19:40:43.841Z] Copy ./auditbeat/build into build/auditbeat/build
[2020-09-30T19:40:43.861Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-21436/src/github.com/elastic/beats/build
[2020-09-30T19:40:43.918Z] WARNING: Unknown parameter(s) found for class type 'hudson.tasks.junit.pipeline.JUnitResultsStep': id,stashedTestReports
[2020-09-30T19:40:43.922Z] Recording test results
[2020-09-30T19:40:44.169Z] None of the test reports contained any result
[2020-09-30T19:40:44.225Z] Stashed 0 file(s)
[2020-09-30T19:40:44.246Z] Archiving artifacts
[2020-09-30T19:40:44.380Z] ‘**/build/TEST*.out’ doesn’t match anything: ‘**’ exists but not ‘**/build/TEST*.out’
[2020-09-30T19:40:44.380Z] No artifacts found that match the file pattern "**/build/TEST*.out". Configuration error?
[2020-09-30T19:40:44.763Z] + python .ci/scripts/search_system_tests.py
[2020-09-30T19:40:44.791Z] [INFO] system-tests=''. If no empty then let's create a tarball
[2020-09-30T19:40:45.212Z] Client: Docker Engine - Community
[2020-09-30T19:40:45.212Z]  Version:           19.03.13
[2020-09-30T19:40:45.212Z]  API version:       1.40
[2020-09-30T19:40:45.212Z]  Go version:        go1.13.15
[2020-09-30T19:40:45.212Z]  Git commit:        4484c46d9d
[2020-09-30T19:40:45.212Z]  Built:             Wed Sep 16 17:02:36 2020
[2020-09-30T19:40:45.212Z]  OS/Arch:           linux/amd64
[2020-09-30T19:40:45.212Z]  Experimental:      false
[2020-09-30T19:40:45.212Z] 
[2020-09-30T19:40:45.212Z] Server: Docker Engine - Community
[2020-09-30T19:40:45.212Z]  Engine:
[2020-09-30T19:40:45.212Z]   Version:          19.03.13
[2020-09-30T19:40:45.212Z]   API version:      1.40 (minimum version 1.12)
[2020-09-30T19:40:45.212Z]   Go version:       go1.13.15
[2020-09-30T19:40:45.212Z]   Git commit:       4484c46d9d
[2020-09-30T19:40:45.212Z]   Built:            Wed Sep 16 17:01:06 2020
[2020-09-30T19:40:45.212Z]   OS/Arch:          linux/amd64
[2020-09-30T19:40:45.212Z]   Experimental:     false
[2020-09-30T19:40:45.212Z]  containerd:
[2020-09-30T19:40:45.212Z]   Version:          1.3.7
[2020-09-30T19:40:45.212Z]   GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
[2020-09-30T19:40:45.212Z]  runc:
[2020-09-30T19:40:45.212Z]   Version:          1.0.0-rc10
[2020-09-30T19:40:45.212Z]   GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
[2020-09-30T19:40:45.212Z]  docker-init:
[2020-09-30T19:40:45.212Z]   Version:          0.18.0
[2020-09-30T19:40:45.212Z]   GitCommit:        fec3683
[2020-09-30T19:40:49.764Z] Stage "Build&Test" skipped due to earlier failure(s)
[2020-09-30T19:40:52.275Z] [INFO] unstashV2: JOB_GCS_BUCKET is set. bucket param got precedency instead.
[2020-09-30T19:40:52.301Z] [INFO] unstashV2: JOB_GCS_CREDENTIALS is set. credentialsId param got precedency instead.
[2020-09-30T19:40:52.384Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs://beats-ci-temp/Beats/beats/PR-21436-1/source/source.tgz
[2020-09-30T19:40:52.403Z] [Google Cloud Storage Plugin] Downloading: Beats/beats/PR-21436-1/source/source.tgz to local path: /var/lib/jenkins/workspace/Beats_beats_PR-21436/source.tgz
[2020-09-30T19:41:05.287Z] + tar --version
[2020-09-30T19:41:05.611Z] + tar -xpf source.tgz
[2020-09-30T19:41:18.219Z] + rm source.tgz
[2020-09-30T19:41:18.313Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-21436/src/github.com/elastic/beats
[2020-09-30T19:41:18.343Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-21436/src/github.com/elastic/beats/uncategorized-1601494844170
[2020-09-30T19:41:18.773Z] + cat
[2020-09-30T19:41:18.773Z] + /usr/local/bin/runbld ./runbld-test-reports --job-name elastic+beats+pull-request
[2020-09-30T19:41:18.773Z] Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[2020-09-30T19:41:25.388Z] runbld>>> runbld started
[2020-09-30T19:41:25.388Z] runbld>>> 1.6.12/f45d832f2ba0aa2722ab4ec1fda8ad140f027f8b
[2020-09-30T19:41:27.312Z] runbld>>> The following profiles matched the job 'elastic+beats+pull-request' in order of occurrence in the config (last value wins).
[2020-09-30T19:41:27.312Z] runbld>>> Matches in the system config:
[2020-09-30T19:41:27.312Z] runbld>>> - Matched ^elastic\+beats
[2020-09-30T19:41:27.312Z] runbld>>> - Matched ^elastic\+beats\+pull-request
[2020-09-30T19:41:28.705Z] runbld>>> Debug logging enabled.
[2020-09-30T19:41:28.705Z] runbld>>> Storing result
[2020-09-30T19:41:28.705Z] runbld>>> Store result: created {:total 2, :successful 2, :failed 0} 1
[2020-09-30T19:41:28.706Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20200930194128-D4E1FCA6
[2020-09-30T19:41:28.706Z] runbld>>> Adding system facts.
[2020-09-30T19:41:30.100Z] runbld>>> Adding vcs info for the latest commit:  f755ba6f38909a7e1cd7d7dbbc0c8c7272689a4b
[2020-09-30T19:41:30.100Z] runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
[2020-09-30T19:41:30.100Z] runbld>>> Adding /usr/lib/jvm/java-8-openjdk-amd64/bin to the path.
[2020-09-30T19:41:30.100Z] + echo 'Processing JUnit reports with runbld...'
[2020-09-30T19:41:30.100Z] Processing JUnit reports with runbld...
[2020-09-30T19:41:30.364Z] runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
[2020-09-30T19:41:30.364Z] runbld>>> DURATION: 28ms
[2020-09-30T19:41:30.364Z] runbld>>> STDOUT: 40 bytes
[2020-09-30T19:41:30.364Z] runbld>>> STDERR: 49 bytes
[2020-09-30T19:41:30.364Z] runbld>>> WRAPPED PROCESS: SUCCESS (0)
[2020-09-30T19:41:30.364Z] runbld>>> Searching for build metadata in /var/lib/jenkins/workspace/Beats_beats_PR-21436
[2020-09-30T19:41:31.756Z] runbld>>> Storing build metadata: 
[2020-09-30T19:41:31.756Z] runbld>>> Adding test report.
[2020-09-30T19:41:31.756Z] runbld>>> Searching for junit test output files with the pattern: TEST-.*\.xml$ in: /var/lib/jenkins/workspace/Beats_beats_PR-21436/src/github.com/elastic/beats
[2020-09-30T19:41:32.333Z] runbld>>> Found 0 test output files
[2020-09-30T19:41:32.333Z] runbld>>> Test output logs contained: Errors: 0 Failures: 0 Tests: 0 Skipped: 0
[2020-09-30T19:41:32.334Z] runbld>>> Storing result
[2020-09-30T19:41:32.595Z] runbld>>> Store result: updated {:total 2, :successful 2, :failed 0} 2
[2020-09-30T19:41:32.595Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20200930194128-D4E1FCA6
[2020-09-30T19:41:32.858Z] runbld>>> Email notification disabled by environment variable.
[2020-09-30T19:41:32.859Z] runbld>>> Slack notification disabled by environment variable.
[2020-09-30T19:41:38.347Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-21436
[2020-09-30T19:41:38.473Z] [INFO] getVaultSecret: Getting secrets
[2020-09-30T19:41:38.562Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-09-30T19:41:39.469Z] + chmod 755 generate-build-data.sh
[2020-09-30T19:41:39.469Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21436/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21436/runs/1 FAILURE 655131
[2020-09-30T19:41:39.469Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21436/runs/1/steps/?limit=10000 -o steps-info.json
[2020-09-30T19:41:40.020Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-21436/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
## :grey_exclamation: Flaky test report
No test was executed to be analysed.





</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_flaky.id-->

blakerouse commented on go.sum:
```diff
@@ -241,6 +241,8 @@ github.com/eapache/queue v1.1.0 h1:YOEu7KNc61ntiQlcEeUIoDTJ2o8mQznoNvUhiigpIqc=
 github.com/eclipse/paho.mqtt.golang v1.2.1-0.20200121105743-0d940dd29fd2 h1:DW6WrARxK5J+o8uAKCiACi5wy9EK1UzrsCpGBPsKHAA=
 github.com/eclipse/paho.mqtt.golang v1.2.1-0.20200121105743-0d940dd29fd2/go.mod h1:H9keYFcgq3Qr5OUJm/JZI/i6U7joQ8SYLhZwfeOo6Ts=
+github.com/elastic/beats v1.3.1 h1:hHzUBHCo3HJHxnRVwa0XlfZoxmP8Rxp7GQ0ZVELGY4A=
+github.com/elastic/beats v7.6.2+incompatible h1:jHdLv83KURaqWUC6f55iMyVP6LYZrgElfeqxKWcskVE=
```
Don't think this should be here.

blakerouse commented on heartbeat/seccomp_linux.go:
```diff
@@ -0,0 +1,35 @@
+			panic(err)
+		}
+	case "arm":
+		// TODO: Figure out how to allow execve here
```
Is arm? Really different here? It is still execve, being is still Linux.

blakerouse commented on x-pack/heartbeat/monitors/browser/synthexec/enrich.go:
```diff
@@ -0,0 +1,104 @@
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// journeyEnricher olds state across received SynthEvents retaining fields
```
s/olds/holds

blakerouse reviewed (commented):
Overall this looks good, just a few minor comments.

Would like to see more testing of the execution of the synthetics node module communicating with the go code.

andrewvc commented:
@blakerouse this is finally ready for real review prior to being merged into master. I've updated the PR description

Some key changes since you last reviewed it:

1. Seccomp rules are fixed based on seccomp profiling, users no longer need to run with seccomp disabled
2. Added more tests for the forked process
3. Fixed the broken field docs
4. This is still experimental, but I'll no longer be pushing directly to this branch, future changes will come with regular PR workflows. By 7.12 FF the goal is to get rid of the custom docker image here and use the main heartbeat docker image for synthetics (see: https://github.com/elastic/beats/issues/22837 )
5. Since we still only want people to run this in the docker image, we guard against accidental use outside of that by checking for the env var `ELASTIC_SYNTHETICS_CAPABLE`, which we only set in that image (yes, this is easy to circumvent, but we want users to realize they are circumventing something before they do so)

andrewvc commented on x-pack/heartbeat/monitors/browser/synthexec/synthexec_test.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func TestRunCmd(t *testing.T) {
+	cmd := exec.Command("go", "run", "./main.go")
```
@blakerouse I hope this test addresses your concern re: covering actual binary execution!

blakerouse commented on x-pack/heartbeat/seccomp_linux.go:
```diff
@@ -0,0 +1,85 @@
+			"with",
+			"write"}
+
+		fmt.Println("SYSCALLS !!!\n")
```
I think you forgot to remove this line.

andrewvc replied:
Ah, good catch!

blakerouse reviewed (approved):
This looks great and excited to see it finally landing.

I am interested in getting this to work under Agent without needing a container. Seems playwright works on Linux, Mac, and Windows. I am sure we can come up with a way to bundle it all, not requiring a special container.

andrewvc commented:
/packaging

elasticmachine commented:
Pinging @elastic/uptime (Team:Uptime)

andrewvc commented:
/packaging test
/test

andrewvc commented:
/test this

andrewvc commented:
/packaging this

andrewvc commented:
/test this

andrewvc commented:
/test this

v1v commented on .ci/heartbeat-synthetics.groovy:
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env groovy
```
Is this particular pipeline required? The current CI pipeline does not use it, so I wonder if the packaging pipeline should be the one to be used instead?

andrewvc replied:
@kuisathaverat would be the best one to ping here since he added that to this PR. However, I do believe it's essential (today) to make the custom synthetics image based off this branch.

kuisathaverat replied:
since the x-pack/heartbeat is build in package pipeline, I do not think we need the groovy file. the heartbeat Docker image will be the new synthetics  docker image. this pipeline is only to change the name to the Docker image to `...-synthetics`
